pyspectral/rsr_reader.py

```diff
@@ -262,6 +216,87 @@ def convert(self):
+        """Set the band names."""
+        self.band_names = h5f.attrs['band_names']
+        self.band_names = [np2str(x) for x in self.band_names]
+
+    def set_instrument(self, h5f):
+        """Set the instrument name."""
+        if self.instrument:
+            return
+
+        try:
+            self.instrument = np2str(h5f.attrs['sensor'])
+        except KeyError:
+            LOG.warning("No sensor name specified in HDF5 file")
+            self.instrument = INSTRUMENTS.get(self.platform_name)
+
+    def set_platform_name(self, h5f):
+        """Set the platform name."""
+        if self.platform_name:
+            return
+
+        try:
+            self.platform_name = np2str(h5f.attrs['platform_name'])
+        except KeyError:
+            LOG.warning("No platform_name in HDF5 file")
+            try:
+                self.platform_name = np2str(h5f.attrs['platform']) + '-' + str(h5f.attrs['sat_number'])
+            except KeyError:
+                LOG.warning(
+                    "Unable to determine platform name from HDF5 file content")
+                self.platform_name = None
+
+        self.platform_name = OSCAR_PLATFORM_NAMES.get(self.platform_name, self.platform_name)
+
+    def get_relative_spectral_responses(self, h5f):
+        """Read the rsr data and add to the object."""
+        no_detectors_message = False
+        for bandname in self.band_names:
+            self.rsr[bandname] = {}
+            try:
+                num_of_det = h5f[bandname].attrs['number_of_detectors']
+            except KeyError:
+                if not no_detectors_message:
+                    LOG.debug("No detectors found - assume only one...")
+                num_of_det = 1
+                no_detectors_message = True
+
+            for i in range(1, num_of_det + 1):
+                dname = 'det-{0:d}'.format(i)
+                self.rsr[bandname][dname] = {}
+                try:
+                    resp = h5f[bandname][dname]['response'][:]
+                except KeyError:
+                    resp = h5f[bandname]['response'][:]
+
+                self.rsr[bandname][dname]['response'] = resp
+
+                try:
+                    wvl = (h5f[bandname][dname]['wavelength'][:] *
+                           h5f[bandname][dname][
+                               'wavelength'].attrs['scale'])
+                except KeyError:
+                    wvl = (h5f[bandname]['wavelength'][:] *
+                           h5f[bandname]['wavelength'].attrs['scale'])
+
+                # The wavelength is given in micro meters!
+                self.rsr[bandname][dname]['wavelength'] = wvl * 1e6
+
+                try:
+                    central_wvl = h5f[bandname][
+                        dname].attrs['central_wavelength']
+                except KeyError:
+                    central_wvl = h5f[bandname].attrs['central_wavelength']
+
+                self.rsr[bandname][dname][
+                    'central_wavelength'] = central_wvl
+
```

Comment:
This function should be refactored, it is too long and cryptic

Reply:
I agree! I will do!